--- vdr/osd.h.orig	2015-04-19 14:18:25.000000000 +0200
+++ vdr/osd.h	2017-02-25 11:56:10.863339207 +0100
@@ -785,7 +785,7 @@
        ///< If there are no dirty pixmaps, or if this is not a true color OSD,
        ///< this function returns NULL.
        ///< The caller must call DestroyPixmap() for the returned pixmap after use.
-//#define DEPRECATED_GETBITMAP
+#define DEPRECATED_GETBITMAP
 #ifdef DEPRECATED_GETBITMAP
 public:
 #endif
