--- vdr-2.3.2-vanilla/menu.c	2016-12-22 13:00:13.000000000 +0200
+++ vdr-2.3.2-editrecording/menu.c	2017-01-05 14:48:24.227929683 +0200
@@ -2516,9 +2516,11 @@
   cMenuEditStrItem *nameItem;
   const char *buttonFolder;
   const char *buttonAction;
-  const char *buttonDeleteMarks;
+  const char *buttonDelete;
   const char *actionCancel;
   const char *doCut;
+  const char *doCopy;
+  bool extraAction;
   int recordingIsInUse;
   void Set(void);
   void SetHelpKeys(void);
@@ -2527,7 +2529,7 @@
   eOSState Folder(void);
   eOSState Action(void);
   eOSState RemoveName(void);
-  eOSState DeleteMarks(void);
+  eOSState Delete(void);
   eOSState ApplyChanges(void);
 public:
   cMenuRecordingEdit(const cRecording *Recording);
@@ -2548,9 +2550,11 @@
   nameItem = NULL;
   buttonFolder = NULL;
   buttonAction = NULL;
-  buttonDeleteMarks = NULL;
+  buttonDelete = NULL;
   actionCancel = NULL;
   doCut = NULL;
+  doCopy = NULL;
+  extraAction = false;
   recordingIsInUse = ruNone;
   Set();
 }
@@ -2582,20 +2586,26 @@
 {
   buttonFolder = !recordingIsInUse ? tr("Button$Folder") : NULL;
   buttonAction = NULL;
-  buttonDeleteMarks = NULL;
+  buttonDelete = NULL;
   actionCancel = NULL;
   doCut = NULL;
+  doCopy = NULL;
   if ((recordingIsInUse & ruCut) != 0)
      buttonAction = actionCancel = ((recordingIsInUse & ruPending) != 0) ? tr("Button$Cancel cutting") : tr("Button$Stop cutting");
   else if ((recordingIsInUse & ruMove) != 0)
      buttonAction = actionCancel = ((recordingIsInUse & ruPending) != 0) ? tr("Button$Cancel moving") : tr("Button$Stop moving");
   else if ((recordingIsInUse & ruCopy) != 0)
      buttonAction = actionCancel = ((recordingIsInUse & ruPending) != 0) ? tr("Button$Cancel copying") : tr("Button$Stop copying");
+  else if (extraAction) {
+     cResumeFile ResumeFile(recording->FileName(), recording->IsPesRecording());
+     buttonAction = doCopy = tr("Button$Copy");
+     buttonDelete = (ResumeFile.Read() != -1) ? tr("Button$Delete resume") : NULL;
+     }
   else if (recording->HasMarks()) {
      buttonAction = doCut = tr("Button$Cut");
-     buttonDeleteMarks = tr("Button$Delete marks");
+     buttonDelete = tr("Button$Delete marks");
      }
-  SetHelp(buttonFolder, buttonAction, buttonDeleteMarks);
+  SetHelp(buttonFolder, buttonAction, buttonDelete, tr("Button$Toggle commands"));
 }
 
 bool cMenuRecordingEdit::RefreshRecording(void)
@@ -2638,6 +2648,25 @@
            Skins.Message(mtError, tr("Error while queueing recording for cutting!"));
         }
      }
+  else if (doCopy) {
+     if (!*name)
+        *name = ' '; // name must not be empty!
+     cString NewName = *folder ? cString::sprintf("%s%c%s", folder, FOLDERDELIMCHAR, name) : name;
+     NewName.CompactChars(FOLDERDELIMCHAR);
+     if (strcmp(NewName, recording->Name())) {
+        cString FromName = cString(ExchangeChars(strdup(recording->Name()), true), true);
+        cString ToName = cString(ExchangeChars(strdup(*NewName), true), true);
+        cString FileName = cString(strreplace(strdup(recording->FileName()), *FromName, *ToName), true);
+        if (access(*FileName, F_OK) != 0 || Interface->Confirm(tr("Edited version already exists - overwrite?"))) {
+           if (MakeDirs(FileName, true) && !RecordingsHandler.Add(ruCopy, recording->FileName(), FileName))
+              Skins.Message(mtError, tr("Error while queueing recording for copying!"));
+           else {
+              LOCK_RECORDINGS_WRITE;
+              Recordings->AddByName(FileName);
+              }
+           }
+        }
+     }
   recordingIsInUse = recording->IsInUse();
   RefreshRecording();
   SetHelpKeys();
@@ -2662,13 +2691,22 @@
   return osContinue;
 }
 
-eOSState cMenuRecordingEdit::DeleteMarks(void)
+eOSState cMenuRecordingEdit::Delete(void)
 {
-  if (buttonDeleteMarks && Interface->Confirm(tr("Delete editing marks for this recording?"))) {
-     if (cMarks::DeleteMarksFile(recording))
+  if (extraAction) {
+     if (buttonDelete && Interface->Confirm(tr("Delete resume for this recording?"))) {
+        cResumeFile ResumeFile(recording->FileName(), recording->IsPesRecording());
+        ResumeFile.Delete();
         SetHelpKeys();
-     else
-        Skins.Message(mtError, tr("Error while deleting editing marks!"));
+        }
+     }
+  else {
+     if (buttonDelete && Interface->Confirm(tr("Delete editing marks for this recording?"))) {
+        if (cMarks::DeleteMarksFile(recording))
+           SetHelpKeys();
+         else
+           Skins.Message(mtError, tr("Error while deleting editing marks!"));
+        }
      }
   return osContinue;
 }
@@ -2728,7 +2766,8 @@
        case k0:      return RemoveName();
        case kRed:    return buttonFolder ? Folder() : osContinue;
        case kGreen:  return buttonAction ? Action() : osContinue;
-       case kYellow: return buttonDeleteMarks ? DeleteMarks() : osContinue;
+       case kYellow: return buttonDelete ? Delete() : osContinue;
+       case kBlue:   extraAction = !extraAction; SetHelpKeys(); return osContinue;
        case kOk:     return !recordingIsInUse ? ApplyChanges() : osBack;
        default: break;
        }
--- vdr-2.3.2-vanilla/po/fi_FI.po	2016-12-23 16:58:03.462451504 +0200
+++ vdr-2.3.2-editrecording/po/fi_FI.po	2017-01-05 14:48:24.227929683 +0200
@@ -794,12 +794,21 @@
 msgid "Button$Stop copying"
 msgstr "Lopeta kopiointi"
 
+msgid "Button$Copy"
+msgstr "Kopioi"
+
+msgid "Button$Delete resume"
+msgstr "Poista paluutiedot"
+
 msgid "Button$Cut"
 msgstr "Leikkaa"
 
 msgid "Button$Delete marks"
 msgstr "Poista merkinnät"
 
+msgid "Button$Toggle commands"
+msgstr "Vaihda komennot"
+
 msgid "Recording vanished!"
 msgstr "Tallenne katosi!"
 
@@ -812,6 +821,12 @@
 msgid "Rename recording to folder name?"
 msgstr "Nimeä tallenne kansion mukaan?"
 
+msgid "Error while queueing recording for copying!"
+msgstr "Tallenteen lisääminen kopiointijonoon epäonnistui!"
+
+msgid "Delete resume for this recording?"
+msgstr "Poistetaanko tallenteen paluutiedot?"
+
 msgid "Delete editing marks for this recording?"
 msgstr "Poistetaanko muokkausmerkinnät tallenteelta?"
 
--- vdr-2.3.2-vanilla/svdrp.c	2016-12-08 12:48:53.000000000 +0200
+++ vdr-2.3.2-editrecording/svdrp.c	2017-01-05 14:48:24.228929683 +0200
@@ -725,6 +725,10 @@
   "    After a CLRE command, no further EPG processing is done for 10\n"
   "    seconds, so that data sent with subsequent PUTE commands doesn't\n"
   "    interfere with data from the broadcasters.",
+  "CPYR <number> <new name>\n"
+  "    Copy the recording with the given number. Before a recording can be\n"
+  "    copied, an LSTR command must have been executed in order to retrieve\n"
+  "    the recording numbers.\n"
   "DELC <number>\n"
   "    Delete channel.",
   "DELR <id>\n"
@@ -951,6 +955,7 @@
   void PrintHelpTopics(const char **hp);
   void CmdCHAN(const char *Option);
   void CmdCLRE(const char *Option);
+  void CmdCPYR(const char *Option);
   void CmdDELC(const char *Option);
   void CmdDELR(const char *Option);
   void CmdDELT(const char *Option);
@@ -1284,6 +1289,57 @@
   return NULL;
 }
 
+void cSVDRPServer::CmdCPYR(const char *Option)
+{
+  if (*Option) {
+     char *opt = strdup(Option);
+     char *num = skipspace(opt);
+     char *option = num;
+     while (*option && !isspace(*option))
+           option++;
+     char c = *option;
+     *option = 0;
+     if (isnumber(num)) {
+        LOCK_RECORDINGS_WRITE;
+        Recordings->SetExplicitModify();
+        if (cRecording *Recording = Recordings->Get(strtol(num, NULL, 10) - 1)) {
+           if (int RecordingInUse = Recording->IsInUse())
+              Reply(550, "%s", *RecordingInUseMessage(RecordingInUse, Option, Recording));
+           else {
+              if (c)
+                 option = skipspace(++option);
+              if (*option) {
+                 cString newName = option;
+                 newName.CompactChars(FOLDERDELIMCHAR);
+                 if (strcmp(newName, Recording->Name())) {
+                    cString fromName = cString(ExchangeChars(strdup(Recording->Name()), true), true);
+                    cString toName = cString(ExchangeChars(strdup(*newName), true), true);
+                    cString fileName = cString(strreplace(strdup(Recording->FileName()), *fromName, *toName), true);
+                    if (MakeDirs(fileName, true) && !RecordingsHandler.Add(ruCopy, Recording->FileName(), fileName)) {
+                       Recordings->AddByName(fileName);
+                       Reply(250, "Recording \"%s\" copied to \"%s\"", Recording->Name(), *newName);
+                       }
+                    else
+                       Reply(554, "Error while copying recording \"%s\" to \"%s\"!", Recording->Name(), *newName);
+                    }
+                 else
+                    Reply(501, "Identical new recording name");
+                 }
+              else
+                 Reply(501, "Missing new recording name");
+              }
+           }
+        else
+           Reply(550, "Recording \"%s\" not found", num);
+        }
+     else
+        Reply(501, "Error in recording number \"%s\"", num);
+     free(opt);
+     }
+  else
+     Reply(501, "Missing recording number");
+}
+
 void cSVDRPServer::CmdDELR(const char *Option)
 {
   if (*Option) {
