diff -Naurp vdr-2.3.7/eit.c vdr-2.3.7-patch/eit.c
--- vdr-2.3.7/eit.c	2017-06-20 09:30:54.000000000 +0300
+++ vdr-2.3.7-patch/eit.c	2017-06-20 10:10:30.000000000 +0300
@@ -24,10 +24,10 @@
 
 class cEIT : public SI::EIT {
 public:
-  cEIT(cSectionSyncerHash &SectionSyncerHash, int Source, u_char Tid, const u_char *Data, int Transponder = 0);
+  cEIT(cSectionSyncerHash &SectionSyncerHash, int Source, u_char Tid, const u_char *Data);
   };
 
-cEIT::cEIT(cSectionSyncerHash &SectionSyncerHash, int Source, u_char Tid, const u_char *Data, int Transponder)
+cEIT::cEIT(cSectionSyncerHash &SectionSyncerHash, int Source, u_char Tid, const u_char *Data)
 :SI::EIT(Data, false)
 {
   if (!CheckCRCAndParse())
@@ -53,7 +53,7 @@ cEIT::cEIT(cSectionSyncerHash &SectionSy
      return;
      }
   tChannelID channelID(Source, getOriginalNetworkId(), getTransportStreamId(), getServiceId());
-  cChannel *Channel = Channels->GetByServiceID(Source, Transponder, getServiceId());
+  cChannel *Channel = Channels->GetByChannelID(channelID, true);
   if (!Channel || EpgHandlers.IgnoreChannel(Channel)) {
      ChannelsStateKey.Remove(false);
      return;
@@ -416,7 +416,7 @@ void cEitFilter::Process(u_short Pid, u_
   switch (Pid) {
     case 0x12: {
          if (Tid >= 0x4E && Tid <= 0x6F)
-            cEIT EIT(sectionSyncerHash, Source(), Tid, Data, Transponder());
+            cEIT EIT(sectionSyncerHash, Source(), Tid, Data);
          }
          break;
     case 0x14: {
