--- osd2web/skin.c
+++ osd2web/skin.c
@@ -79,7 +79,9 @@
 class cSkinOsd2WebDisplayMenu : public cSkinDisplayMenu
 {
    public:
-
+#ifdef DEPRECATED_SKIN_SETITEMEVENT
+      using cSkinDisplayMenu::SetItemEvent;
+#endif
       cSkinOsd2WebDisplayMenu();
       virtual ~cSkinOsd2WebDisplayMenu();
 
@@ -91,7 +93,7 @@
       virtual void SetButtons(const char* Red, const char* Green = NULL, const char* Yellow = NULL, const char* Blue = NULL);
       virtual void SetMessage(eMessageType Type, const char* Text);
       virtual void SetItem(const char* Text, int Index, bool Current, bool Selectable);
-      virtual bool SetItemEvent(const cEvent* Event, int Index, bool Current, bool Selectable, const cChannel *Channel, bool WithDate, eTimerMatch TimerMatch);
+      virtual bool SetItemEvent(const cEvent* Event, int Index, bool Current, bool Selectable, const cChannel *Channel, bool WithDate, eTimerMatch TimerMatch, bool TimerActive);
       virtual bool SetItemChannel(const cChannel* Channel, int Index, bool Current, bool Selectable, bool WithProvider);
       virtual bool SetItemRecording(const cRecording* Recording, int Index, bool Current, bool Selectable, int Level, int Total, int New);
       virtual bool SetItemTimer(const cTimer* Timer, int Index, bool Current, bool Selectable);
@@ -219,7 +221,7 @@
 
 bool cSkinOsd2WebDisplayMenu::SetItemEvent(const cEvent* Event, int Index, bool Current,
                                            bool Selectable, const cChannel *Channel,
-                                           bool WithDate, eTimerMatch TimerMatch)
+                                           bool WithDate, eTimerMatch TimerMatch, bool TimerActive)
 {
    tell(4, "DEB: Skin:cSkinOsd2WebDisplayMenu::SetItemEvent()");
 
