diff --git a/drivers/amlogic/amports/vvc1.c b/drivers/amlogic/amports/vvc1.c
index 0dd5f1a..92abac1 100644
--- a/drivers/amlogic/amports/vvc1.c
+++ b/drivers/amlogic/amports/vvc1.c
@@ -286,7 +286,7 @@ static irqreturn_t vvc1_isr(int irq, void *dev_id)
 
 
                 repeat_count = READ_VREG(VC1_REPEAT_COUNT);
-                buffer_index = ((reg & 0x7) - 1) & 3;
+                buffer_index = reg & 0x7;
                 picture_type = (reg >> 3) & 7;
 
 		if (pts_by_offset) {
@@ -305,10 +305,6 @@ static irqreturn_t vvc1_isr(int irq, void *dev_id)
 			}
 		}
 
-		repeat_count = READ_VREG(VC1_REPEAT_COUNT);
-		buffer_index = reg & 0x7;
-		picture_type = (reg >> 3) & 7;
-
 		if (buffer_index >= DECODE_BUFFER_NUM_MAX) {
 			pr_info("fatal error, invalid buffer index.");
 			return IRQ_HANDLED;
