--- a/xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp 2017-04-21 14:42:16.371525222 +0200
+++ b/xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp 2017-04-21 14:43:32.295525222 +0200
@@ -1287,7 +12887,7 @@
     double err = fmod(renderPts - nextFramePts, frametime);
     m_clockSync.m_error += err;
     m_clockSync.m_errCount ++;
-    if (m_clockSync.m_errCount > 30)
+    if (m_clockSync.m_errCount > 120)
     {
       double average = m_clockSync.m_error / m_clockSync.m_errCount;
       m_clockSync.m_syncOffset = average;