diff -Naurp media_build_cc-20170408/linux/drivers/media/dvb-frontends/si2183.c media_build_cc-20170408-patch/linux/drivers/media/dvb-frontends/si2183.c
--- media_build_cc-20170408/linux/drivers/media/dvb-frontends/si2183.c	2017-04-08 08:23:38.000000000 +0300
+++ media_build_cc-20170408-patch/linux/drivers/media/dvb-frontends/si2183.c	2017-04-14 17:44:13.000000000 +0300
@@ -81,6 +81,7 @@ struct si2183_dev {
 	bool ter_agc_inv;
 	bool sat_agc_inv;
 
+	u8 agc_mode;
 	struct si_base *base;
 
 	void (*RF_switch)(struct i2c_adapter * i2c,u8 rf_in,u8 flag);
@@ -427,10 +428,19 @@ static int si2183_set_dvbc(struct dvb_fr
 {
 	struct dtv_frontend_properties *c = &fe->dtv_property_cache;
 	struct i2c_client *client = fe->demodulator_priv;
+	struct si2183_dev *dev = i2c_get_clientdata(client);
 	struct si2183_cmd cmd;
 	int ret;
 	u16 prop;
 
+	memcpy(cmd.args, "\x89\x41\x06\x12\x0\x0", 6);
+	cmd.args[1]= (dev->agc_mode &0x07)<<4 |0x1;  
+ 	cmd.wlen = 6;
+ 	cmd.rlen = 3;
+ 	ret = si2183_cmd_execute(client, &cmd);
+ 	if(ret){
+ 		printk("Ter AGC set error !");
+ 	}
 	/* dvb-c mode */
 	prop = 0x38;
 	ret = si2183_set_prop(client, SI2183_PROP_MODE, &prop);
@@ -551,11 +561,22 @@ static int si2183_set_dvbs(struct dvb_fr
 {
 	struct dtv_frontend_properties *c = &fe->dtv_property_cache;
 	struct i2c_client *client = fe->demodulator_priv;
+	struct si2183_dev *dev = i2c_get_clientdata(client);
 	struct si2183_cmd cmd;
 	int ret;
 	u16 prop;
 	u32 pls_mode, pls_code;
-	
+
+	/*set SAT agc*/
+	memcpy(cmd.args, "\x8a\x1d\x12\x0\x0\x0", 6);
+	cmd.args[1]= dev->agc_mode|0x18;  
+ 	cmd.wlen = 6;
+ 	cmd.rlen = 3;
+ 	ret = si2183_cmd_execute(client, &cmd);
+ 	if(ret){
+ 		printk("AGC set error !");
+ 	}
+
 	/* set mode */
 	prop = 0x8;
 	switch (c->delivery_system) {
@@ -627,10 +648,20 @@ static int si2183_set_dvbt(struct dvb_fr
 {
 	struct dtv_frontend_properties *c = &fe->dtv_property_cache;
 	struct i2c_client *client = fe->demodulator_priv;
+	struct si2183_dev *dev = i2c_get_clientdata(client);
 	struct si2183_cmd cmd;
 	int ret;
 	u16 prop;
 
+	memcpy(cmd.args, "\x89\x41\x06\x12\x0\x0", 6);
+	cmd.args[1]= (dev->agc_mode &0x07)<<4 |0x1;  
+ 	cmd.wlen = 6;
+ 	cmd.rlen = 3;
+ 	ret = si2183_cmd_execute(client, &cmd);
+ 	if(ret){
+ 		printk("Ter AGC set error !");
+ 	}
+
 	if (c->bandwidth_hz == 0) {
 		return -EINVAL;
 	} else if (c->bandwidth_hz <= 2000000)
@@ -696,9 +727,19 @@ static int si2183_set_isdbt(struct dvb_f
 {
 	struct dtv_frontend_properties *c = &fe->dtv_property_cache;
 	struct i2c_client *client = fe->demodulator_priv;
+	struct si2183_dev *dev = i2c_get_clientdata(client);
+	struct si2183_cmd cmd;
 	int ret;
 	u16 prop;
 
+	memcpy(cmd.args, "\x89\x41\x06\x12\x0\x0", 6);
+	cmd.args[1]= (dev->agc_mode &0x07)<<4 |0x1; 
+ 	cmd.wlen = 6;
+ 	cmd.rlen = 3;
+ 	ret = si2183_cmd_execute(client, &cmd);
+ 	if(ret){
+ 		printk("Ter AGC set error !");
+ 	}
 	if (c->bandwidth_hz == 0) {
 		return -EINVAL;
 	} else if (c->bandwidth_hz <= 2000000)
@@ -1520,6 +1561,7 @@ static int si2183_probe(struct i2c_clien
 	dev->ts_mode = config->ts_mode;
 	dev->ts_clock_inv = config->ts_clock_inv;
 	dev->ts_clock_gapped = config->ts_clock_gapped;
+	dev->agc_mode = config->agc_mode;
 	dev->fef_pin = config->fef_pin;
 	dev->fef_inv = config->fef_inv;
 	dev->agc_pin = config->agc_pin;
diff -Naurp media_build_cc-20170408/linux/drivers/media/dvb-frontends/si2183.h media_build_cc-20170408-patch/linux/drivers/media/dvb-frontends/si2183.h
--- media_build_cc-20170408/linux/drivers/media/dvb-frontends/si2183.h	2017-04-08 08:23:38.000000000 +0300
+++ media_build_cc-20170408-patch/linux/drivers/media/dvb-frontends/si2183.h	2017-04-14 17:45:05.000000000 +0300
@@ -43,6 +43,8 @@ struct si2183_config {
 
 	/* TS clock gapped */
 	bool ts_clock_gapped;
+	/*agc*/
+	u8 agc_mode;
 
 	/* Tuner control pins */
 #define SI2183_MP_NOT_USED	1
diff -Naurp media_build_cc-20170408/linux/drivers/media/usb/dvb-usb/Kconfig media_build_cc-20170408-patch/linux/drivers/media/usb/dvb-usb/Kconfig
--- media_build_cc-20170408/linux/drivers/media/usb/dvb-usb/Kconfig	2017-04-08 08:23:38.000000000 +0300
+++ media_build_cc-20170408-patch/linux/drivers/media/usb/dvb-usb/Kconfig	2017-04-14 17:29:39.000000000 +0300
@@ -431,6 +431,15 @@ config DVB_USB_TBS5520
 	help
 	  Say Y here to support the Turbosight TBS5520 USB2 DVB-T/T2/C/S/S2 device
 
+config DVB_USB_TBS5520se
+	tristate "Turbosight TBS5520se support"
+	depends on DVB_USB
+	select DVB_SI2183 if MEDIA_SUBDRV_AUTOSELECT
+	select MEDIA_TUNER_SI2157 if MEDIA_SUBDRV_AUTOSELECT
+	select MEDIA_TUNER_AV201X if MEDIA_SUBDRV_AUTOSELECT
+	help
+	  Say Y here to support the Turbosight TBS5520se USB2 DVB-T/T2/C/C2/S/S2/S2x device
+
 config DVB_USB_TBS5927
 	tristate "TurboSight TBS5927 DVB-S2 USB2.0 support"
 	depends on DVB_USB
diff -Naurp media_build_cc-20170408/linux/drivers/media/usb/dvb-usb/Makefile media_build_cc-20170408-patch/linux/drivers/media/usb/dvb-usb/Makefile
--- media_build_cc-20170408/linux/drivers/media/usb/dvb-usb/Makefile	2017-04-08 08:23:38.000000000 +0300
+++ media_build_cc-20170408-patch/linux/drivers/media/usb/dvb-usb/Makefile	2017-04-14 17:30:06.000000000 +0300
@@ -112,6 +112,9 @@ obj-$(CONFIG_DVB_USB_TBS5881) += dvb-usb
 dvb-usb-tbs5520-objs := tbs5520.o
 obj-$(CONFIG_DVB_USB_TBS5520) += dvb-usb-tbs5520.o
 
+dvb-usb-tbs5520se-objs = tbs5520se.o
+obj-$(CONFIG_DVB_USB_TBS5520se) += dvb-usb-tbs5520se.o
+
 dvb-usb-tbs5927-objs = tbs5927.o
 obj-$(CONFIG_DVB_USB_TBS5927) += dvb-usb-tbs5927.o
 
